src/internal/obj/local_client.go

```diff
@@ -1,126 +1,149 @@
-		if err := file.Close(); retErr == nil {
-			retErr = err
-		}
-	}()
-	_, err = io.Copy(w, file)
+	defer f.Close()
+	_, err = io.Copy(w, f)
 	return err
 }
 
-func (c *localClient) Delete(_ context.Context, path string) (retErr error) {
-	defer func() { retErr = c.transformError(retErr, path) }()
-	return os.Remove(c.normPath(path))
+func (c *fsClient) Delete(ctx context.Context, name string) error {
+	err := os.Remove(c.finalPathFor(name))
+	if os.IsNotExist(err) {
+		err = nil
+	}
+	return err
+}
+
+func (c *fsClient) Exists(ctx context.Context, name string) (bool, error) {
+	_, err := os.Stat(c.finalPathFor(name))
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
 }
 
-func (c *localClient) Walk(_ context.Context, dir string, walkFn func(name string) error) error {
-	dir = c.normPath(dir)
-	fi, _ := os.Stat(dir)
-	prefix := ""
-	if fi == nil || !fi.IsDir() {
-		dir, prefix = filepath.Split(dir)
-	}
-	err := filepath.Walk(dir, func(path string, fileInfo os.FileInfo, err error) error {
+func (c *fsClient) Walk(ctx context.Context, prefix string, cb func(string) error) error {
+	dirEnts, err := os.ReadDir(filepath.Join(c.rootDir, "objects"))
+	if err != nil {
+		return err
+	}
+	enc := base64.URLEncoding
+	for _, dirEnt := range dirEnts {
+		// TODO: There is a better way to do this (encode the prefix rather than decode all and filter)
+		// but I'm not really sure how to do that with the filesystem API.
+		// We would need to seek to a certain dir entry.
+		name, err := enc.DecodeString(dirEnt.Name())
 		if err != nil {
-			if os.IsNotExist(err) {
-				return nil
-			}
-			return err
+			return errors.Wrapf(err, "parsing object name")
 		}
-		if fileInfo.IsDir() {
-			return nil
-		}
-		relPath, _ := filepath.Rel(c.root, path)
-		if !strings.HasPrefix(filepath.Base(relPath), prefix) {
-			return nil
+		if bytes.HasPrefix(name, []byte(prefix)) {
+			if err := cb(string(name)); err != nil {
+				return err
+			}
 		}
-		return walkFn(relPath)
-	})
-	return err
+	}
+	return nil
 }
 
-func (c *localClient) Exists(ctx context.Context, path string) (bool, error) {
-	_, err := os.Stat(c.normPath(path))
-	if err != nil {
-		if os.IsNotExist(err) {
-			err = nil
-		}
-		return false, err
+func (c *fsClient) BucketURL() string {
+	return fmt.Sprintf("local://%s/", strings.ReplaceAll(filepath.ToSlash(c.rootDir), "/", "."))
+}
+
+func (c *fsClient) stagingPathFor(name string) string {
+	return filepath.Join(c.rootDir, "staging", uuid.NewWithoutDashes())
+}
+
+func (c *fsClient) finalPathFor(name string) string {
+	enc := base64.URLEncoding
+	return filepath.Join(c.rootDir, "objects", enc.EncodeToString([]byte(name)))
+}
+
+func (c *fsClient) init() error {
+	if err := os.RemoveAll(filepath.Join(c.rootDir, "staging")); err != nil {
+		return err
 	}
-	tracing.TagAnySpan(ctx, "err", err)
-	return true, nil
+	if err := os.MkdirAll(filepath.Join(c.rootDir, "staging"), 0o755); err != nil {
```

Comment:
Why is the extra `o` necessary in the file mode?

Reply:
The convention that numbers starting with 0 are interpreted as octal is ancient tribal knowledge that I don't expect everyone to have.  The `o` makes it obvious.

Reply:
Yeah, but I don't think this is what people normally do in Go for setting the file mode. I normally see `0...`.